File: app/src/main/java/com/chefvoice/app/util/StepTimers.kt
package com.chefvoice.app.util

/**
 * A cooking duration found verbatim in a finished method step.
 *
 * [label] is the text exactly as the chef said it, so a timer offered for
 * "simmer for 10 to 12 minutes" reads "10 to 12 minutes" rather than a number the
 * chef never used.
 */
data class StepTimer(
    val label: String,
    val totalSeconds: Int,
    val startIndex: Int
)

/**
 * Finds the durations already present in a method step so the cook-along screen can
 * offer them as timers.
 *
 * This is a **read-only extractor over finished step text**. It is deliberately not
 * part of `CookingSessionParser`: it never edits a step, never contributes to what is
 * saved, and running it or not running it cannot change a single row of
 * `shared/golden-cooking-corpus.tsv`. The parser's own duration handling stays where
 * it is and is not reused here, because that code sums one duration per step for the
 * prep/cook estimate, while a timer needs every duration in the step along with where
 * in the text it sits.
 *
 * It is as literal as the rest of the pipeline. A duration is only recognised when
 * the chef actually stated a number and a time unit; nothing is inferred from a verb,
 * and an unrecognised phrase simply yields no timer rather than a guessed one. Being
 * wrong here means a kitchen timer that ends at the wrong moment, so the bar for
 * matching is high and the failure mode is always "no timer offered".
 */
object StepTimers {
    private val numberWords = mapOf(
        "one" to 1.0, "two" to 2.0, "three" to 3.0, "four" to 4.0, "five" to 5.0,
        "six" to 6.0, "seven" to 7.0, "eight" to 8.0, "nine" to 9.0, "ten" to 10.0,
        "eleven" to 11.0, "twelve" to 12.0, "thirteen" to 13.0, "fourteen" to 14.0,
        "fifteen" to 15.0, "sixteen" to 16.0, "seventeen" to 17.0, "eighteen" to 18.0,
        "nineteen" to 19.0, "twenty" to 20.0, "thirty" to 30.0, "forty" to 40.0,
        "fifty" to 50.0, "sixty" to 60.0, "ninety" to 90.0,
        "half" to 0.5, "quarter" to 0.25
    )

    private const val NUMBER = "\\d+(?:\\.\\d+)?(?:\\s*/\\s*\\d+)?|one|two|three|four|five|six|seven|" +
        "eight|nine|ten|eleven|twelve|thirteen|fourteen|fifteen|sixteen|seventeen|eighteen|nineteen|" +
        "twenty|thirty|forty|fifty|sixty|ninety|half|quarter"

    /**
     * Matches "20 minutes", "10 to 12 minutes", "1 1/2 hours", "half an hour".
     *
     * The optional second number is what makes a range work. A chef who says a range
     * is describing a window, not one instant, and the timer uses the **lower** bound
     * (see [rangeUsesLowerBound]) so it goes off while there is still something to
     * check rather than after the food is past it.
     */
    private val durationPattern = Regex(
        "(?i)\\b(" + NUMBER + ")" +
            "(?:\\s+(\\d+\\s*/\\s*\\d+))?" +
            "(?:\\s*(?:-|–|to|or)\\s*(" + NUMBER + "))?" +
            "\\s*(?:an?\\s+)?" +
            "(seconds?|secs?|minutes?|mins?|hours?|hrs?)\\b"
    )

    /**
     * A range timer counts the lower bound. "Simmer 10 to 12 minutes" should call the
     * chef back at 10 so they can look at the pan, not at 12 when the decision has
     * already been made for them.
     */
    const val rangeUsesLowerBound = true

    /** Every duration in [step], in the order the chef said them. */
    fun timersIn(step: String): List<StepTimer> {
        if (step.isBlank()) return emptyList()
        return durationPattern.findAll(step).mapNotNull { match ->
            val whole = match.groupValues[1]
            val fraction = match.groupValues[2]
            val rangeEnd = match.groupValues[3]
            val unit = match.groupValues[4]

            val base = amountOf(whole) ?: return@mapNotNull null
            // "1 1/2 hours" -- a whole number followed by a bare fraction.
            val amount = if (fraction.isNotBlank()) base + (amountOf(fraction) ?: 0.0) else base
            // A range keeps the lower of the two, which for a well-formed range is
            // the first. A reversed range ("12 to 10") is still read as lower-first
            // rather than reordered, because reordering would be interpreting.
            val effective = if (rangeEnd.isNotBlank() && !rangeUsesLowerBound) {
                amountOf(rangeEnd) ?: amount
            } else {
                amount
            }

            val seconds = secondsFor(effective, unit) ?: return@mapNotNull null
            if (seconds <= 0) return@mapNotNull null
            // Nothing in a kitchen is usefully timed beyond a day, and a match that
            // long is far more likely to be a misread number than a real instruction.
            if (seconds > 24 * 60 * 60) return@mapNotNull null

            StepTimer(
                label = match.value.trim(),
                totalSeconds = seconds,
                startIndex = match.range.first
            )
        }.toList()
    }

    /** The first duration in [step], or null when the chef stated none. */
    fun firstTimerIn(step: String): StepTimer? = timersIn(step).firstOrNull()

    private fun amountOf(token: String): Double? {
        val clean = token.trim().lowercase()
        if (clean.isEmpty()) return null
        numberWords[clean]?.let { return it }
        // "1/2", "3 / 4"
        if (clean.contains('/')) {
            val parts = clean.split('/')
            if (parts.size != 2) return null
            val numerator = parts[0].trim().toDoubleOrNull() ?: return null
            val denominator = parts[1].trim().toDoubleOrNull() ?: return null
            if (denominator == 0.0) return null
            return numerator / denominator
        }
        return clean.toDoubleOrNull()
    }

    private fun secondsFor(amount: Double, unit: String): Int? {
        val lower = unit.lowercase()
        val multiplier = when {
            lower.startsWith("sec") -> 1.0
            lower.startsWith("min") -> 60.0
            lower.startsWith("hour") || lower.startsWith("hr") -> 3600.0
            else -> return null
        }
        val seconds = amount * multiplier
        if (seconds.isNaN() || seconds.isInfinite()) return null
        return Math.round(seconds).toInt()
    }

    /** "1:05:00", "20:00", "0:45" -- how a running timer reads on screen. */
    fun formatClock(totalSeconds: Int): String {
        val safe = totalSeconds.coerceAtLeast(0)
        val hours = safe / 3600
        val minutes = (safe % 3600) / 60
        val seconds = safe % 60
        return if (hours > 0) "%d:%02d:%02d".format(hours, minutes, seconds)
        else "%d:%02d".format(minutes, seconds)
    }
}

File: web/js/step-timers.js
// Port of app/.../util/StepTimers.kt. Keep the two in step.
//
// Finds the durations already present in a finished method step so the cook-along screen
// can offer them as timers. A read-only extractor: it never edits a step and is not part of
// the cooking parser, so it cannot change a golden-corpus row. A duration is only
// recognised when the chef stated a number and a time unit; anything else yields no timer
// rather than a guess, because a wrong timer calls the chef back at the wrong moment.

const numberWords = {
  one: 1, two: 2, three: 3, four: 4, five: 5, six: 6, seven: 7, eight: 8, nine: 9, ten: 10,
  eleven: 11, twelve: 12, thirteen: 13, fourteen: 14, fifteen: 15, sixteen: 16,
  seventeen: 17, eighteen: 18, nineteen: 19, twenty: 20, thirty: 30, forty: 40,
  fifty: 50, sixty: 60, ninety: 90, half: 0.5, quarter: 0.25
};

const NUMBER = '\\d+(?:\\.\\d+)?(?:\\s*/\\s*\\d+)?|one|two|three|four|five|six|seven|' +
  'eight|nine|ten|eleven|twelve|thirteen|fourteen|fifteen|sixteen|seventeen|eighteen|nineteen|' +
  'twenty|thirty|forty|fifty|sixty|ninety|half|quarter';

const durationSource = '\\b(' + NUMBER + ')' +
  '(?:\\s+(\\d+\\s*/\\s*\\d+))?' +
  '(?:\\s*(?:-|–|to|or)\\s*(' + NUMBER + '))?' +
  '\\s*(?:an?\\s+)?' +
  '(seconds?|secs?|minutes?|mins?|hours?|hrs?)\\b';

/** A range times the lower bound, so the chef is called back while there is still a decision. */
export const rangeUsesLowerBound = true;

function amountOf(token) {
  const clean = String(token || '').trim().toLowerCase();
  if (!clean) return null;
  if (Object.prototype.hasOwnProperty.call(numberWords, clean)) return numberWords[clean];
  if (clean.includes('/')) {
    const parts = clean.split('/');
    if (parts.length !== 2) return null;
    const numerator = Number(parts[0].trim());
    const denominator = Number(parts[1].trim());
    if (!Number.isFinite(numerator) || !Number.isFinite(denominator) || denominator === 0) return null;
    return numerator / denominator;
  }
  const value = Number(clean);
  return Number.isFinite(value) ? value : null;
}

function secondsFor(amount, unit) {
  const lower = unit.toLowerCase();
  let multiplier;
  if (lower.startsWith('sec')) multiplier = 1;
  else if (lower.startsWith('min')) multiplier = 60;
  else if (lower.startsWith('hour') || lower.startsWith('hr')) multiplier = 3600;
  else return null;
  const seconds = amount * multiplier;
  return Number.isFinite(seconds) ? Math.round(seconds) : null;
}

/** Every duration in `step`, in the order the chef said them: {label, totalSeconds, startIndex}. */
export function timersIn(step) {
  if (!step || !String(step).trim()) return [];
  const out = [];
  for (const match of String(step).matchAll(new RegExp(durationSource, 'gi'))) {
    const base = amountOf(match[1]);
    if (base === null) continue;
    // "1 1/2 hours" -- a whole number followed by a bare fraction.
    const amount = match[2] ? base + (amountOf(match[2]) ?? 0) : base;
    const effective = match[3] && !rangeUsesLowerBound ? (amountOf(match[3]) ?? amount) : amount;
    const seconds = secondsFor(effective, match[4]);
    if (seconds === null || seconds <= 0) continue;
    // Nothing in a kitchen is usefully timed beyond a day; a match that long is a misread.
    if (seconds > 24 * 60 * 60) continue;
    out.push({ label: match[0].trim(), totalSeconds: seconds, startIndex: match.index });
  }
  return out;
}

export function firstTimerIn(step) {
  return timersIn(step)[0] ?? null;
}

/** "1:05:00", "20:00", "0:45" -- how a running timer reads on screen. */
export function formatClock(totalSeconds) {
  const safe = Math.max(0, Math.trunc(totalSeconds));
  const hours = Math.floor(safe / 3600);
  const minutes = Math.floor((safe % 3600) / 60);
  const seconds = safe % 60;
  const two = n => String(n).padStart(2, '0');
  return hours > 0 ? `${hours}:${two(minutes)}:${two(seconds)}` : `${minutes}:${two(seconds)}`;
}
